### PR TITLE
[checker] Printer cleanup.

### LINKED ISSUE #6799: coqchk is missing many newlines
#### Version

8.8+alpha

#### Operating system

Linux

#### Description of the problem

The output of `coqchk` is missing newlines in many places.  Most egregiously, the entire output is missing a final newline, causing it to run into the subsequent terminal prompt or the next makefile error.  Additionally, newlines are missing between listing arguments of bad applications.  See, for example, #6798 (search for `arg 2=` or `arg 3=` or `arg 4=`).

@ejgallego Comments at https://github.com/coq/coq/issues/6797#issuecomment-367181197 that the code seems to emit a newline, and this does seem to be the case, (though I'm confused why the `CantApplyBadType` case in `explain_exn` seems to open more boxes than it closes, and I'm unsure whether or not this is related)

### PR DESCRIPTION
**Pull request by ejgallego**
Makes printing rules more explicit and should close #6799.

### CODE REVIEW

File: checker/checker.ml [hunk 1]
```diff
@@ -267,26 +267,26 @@ let explain_exn = function
...
-            (ctx.Environ.env_stratification.Environ.env_universes));
-          str "\nCantApplyBadType at argument " ++ int n
+        (* This mix of printf / pp was here before... *)
+        let fmt = Format.std_formatter in
+        let open Format in
+        let open Print in
+        fprintf fmt "====== ill-typed term ====@\n";
+        fprintf fmt "@[<hov 2>application head=@ %a@]@\n" print_pure_constr hd;
+        fprintf fmt "@[<hov 2>head type=@ %a@]@\n" print_pure_constr hdty;
+        let pp_arg fmt (i,(t,ty)) = fprintf fmt "@[<hov 2>arg %d=@ %a@ type=@ %a" (i+1)
```
Comment (JasonGross) [inline checker/checker.ml]: This is missing a box closing between arguments, right?

**Review by JasonGross — CHANGES_REQUESTED** [review]
Thanks!  Would you be willing to paste the output of this on the example in #6798?

File: checker/checker.ml (hunk 1 above)
Comment (ejgallego) [reply checker/checker.ml]: Well-spotted, thanks.

**Comment by ejgallego** [comment]
> Thanks! Would you be willing to paste the output of this on the example in #6798?

This PR is meant to produce byte-identical output; what would you like to change? Maybe informing the `Format` system of wider margins?

**Comment by JasonGross** [comment]
I would like there to be whitespace before `arg` in `arg n=`.  In cases where the argument value does not fit on a single line (or possibly in all cases if that's easier), I'd like the whitespace before `arg` to be a newline.

**Comment by ejgallego** [comment]
Printing is always tricky, I've tweak it a bit to see how it works. We should open a bug thou so coq detects and sets:
```
  Format.(pp_set_margin std_formatter 140);
  Format.(pp_set_max_indent std_formatter 138);
```
automatically on terminals.

**Comment by maximedenes** [comment]
@ejgallego were @JasonGross ' comments addressed? If yes, you may want to dismiss the review.

**Comment by ejgallego** [comment]
@JasonGross can you confirm the current output is OK for you ?

**Comment by JasonGross** [comment]
I haven't had a chance to build and test this, and don't have the time right now, but if you paste the output here on the example from #6798, I can look over it and confirm.

**Comment by ejgallego** [comment]
@JasonGross here it is:
```
====== ill-typed term ====
application head= Cons((k.ap),Var(1) Var(1))
head type=
  (A:Type(Var(1)))
   (B:Type(Var(1)))
    (f:(#2->#2))
     (x:#3)
      (y:#4)
       (p:(Ind((k.paths),0,Var(1)) #5#2#1))
        (Ind((k.paths),0,Var(1)) #5(#4 #3)(#4 #2))
arguments:
arg 1=
 (Ind((k.paths),0,Var(1))
  #4
   (Cons((k.transport),Var(2) Var(1))
    Type(Var(1))[X:Type(Var(1))]#1#5#4
     (Cons((k.path_universe),Var(1) Var(1) Var(2))
      #7#5#4
       Proj((k.equiv_fun),
       (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
       Proj((k.equiv_isequiv),
       (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1)))#2)
   (Proj((k.equiv_fun),
    (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
    #2))
type= Type(Var(1))

arg 2=
 (Ind((k.paths),0,Var(1))
  #5
   (Cons((k.transport),Var(2) Var(1))
    Type(Var(1))[x:Type(Var(1))]#1#4#5
     (Cons((k.inverse),Var(2))
      Type(Var(1))#5#4
       (Cons((k.path_universe),Var(1) Var(1) Var(2))
        #7#5#4
         Proj((k.equiv_fun),
         (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
         (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1)))
     (Cons((k.transport),Var(2) Var(1))
      Type(Var(1))[x:Type(Var(1))]#1#5#4
       (Cons((k.path_universe),Var(1) Var(1) Var(2))
        #7#5#4
         Proj((k.equiv_fun),
         (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
         (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1))
       #2))
   (Cons((k.transport),Var(2) Var(1))
    Type(Var(1))[x:Type(Var(1))]#1#4#5
     (Cons((k.inverse),Var(2))
      Type(Var(1))#5#4
       (Cons((k.path_universe),Var(1) Var(1) Var(2))
        #7#5#4
         Proj((k.equiv_fun),
         (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
         Proj((k.equiv_isequiv),
         (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))))
     (Proj((k.equiv_fun),
      (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
      #2)))
type= Type(Var(1))

arg 3=
 (Cons((k.ap),Var(1) Var(1))
  #4#5
   (Cons((k.transport),Var(2) Var(1))
    Type(Var(1))[x:Type(Var(1))]#1#4#5
     (Cons((k.inverse),Var(2))
      Type(Var(1))#5#4
       (Cons((k.path_universe),Var(1) Var(1) Var(2))
        #7#5#4
         Proj((k.equiv_fun),
         (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
         (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1))))
   (Cons((k.transport),Var(2) Var(1))
    Type(Var(1))[x:Type(Var(1))]#1#5#4
     (Cons((k.path_universe),Var(1) Var(1) Var(2))
      #7#5#4
       Proj((k.equiv_fun),
       (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
       (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1))
     #2)
   (Cons((k.transport),Var(2) Var(1)) Type(Var(1))[x:Type(Var(1))]#1#5#4#1#2))
type=
 (p:(Ind((k.paths),0,Var(1))
   #4
    (Cons((k.transport),Var(2) Var(1))
     Type(Var(1))[x:Type(Var(1))]#1#5#4
      (Cons((k.path_universe),Var(1) Var(1) Var(2))
       #7#5#4
        Proj((k.equiv_fun),
        (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
        (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1))
      #2)
    (Cons((k.transport),Var(2) Var(1))
     Type(Var(1))[x:Type(Var(1))]#1#5#4#1#2)))
  (Ind((k.paths),0,Var(1))
   #6
    ((Cons((k.transport),Var(2) Var(1))
      Type(Var(1))[x:Type(Var(1))]#1#5#6
       (Cons((k.inverse),Var(2))
        Type(Var(1))#6#5
         (Cons((k.path_universe),Var(1) Var(1) Var(2))
          #8#6#5
           Proj((k.equiv_fun),
           (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #6#5#2))
           (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #6#5#2)))
       )
     (Cons((k.transport),Var(2) Var(1))
      Type(Var(1))[x:Type(Var(1))]#1#6#5
       (Cons((k.path_universe),Var(1) Var(1) Var(2))
        #8#6#5
         Proj((k.equiv_fun),
         (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #6#5#2))
         (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #6#5#2))
       #3))
    ((Cons((k.transport),Var(2) Var(1))
      Type(Var(1))[x:Type(Var(1))]#1#5#6
       (Cons((k.inverse),Var(2))
        Type(Var(1))#6#5
         (Cons((k.path_universe),Var(1) Var(1) Var(2))
          #8#6#5
           Proj((k.equiv_fun),
           (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #6#5#2))
           (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #6#5#2)))
       )
     (Cons((k.transport),Var(2) Var(1))
      Type(Var(1))[x:Type(Var(1))]#1#6#5#2#3)))

arg 4=
 (Cons((k.transport_path_universe),Var(0) Var(1) Var(2) Var(1))
  #7#5#4
   Proj((k.equiv_fun),
   (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
   Proj((k.equiv_isequiv),
   (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))#2)
type=
 (Ind((k.paths),0,Var(1))
  #4
   (Cons((k.transport),Var(2) Var(1))
    Type(Var(1))[X:Type(Var(1))]#1#5#4
     (Cons((k.path_universe),Var(1) Var(1) Var(2))
      #7#5#4
       Proj((k.equiv_fun),
       (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
       Proj((k.equiv_isequiv),
       (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1)))#2)
   (Proj((k.equiv_fun),
    (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
    #2))

arg 5=
 (Cons((k.ap),Var(2) Var(1))
  (Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)#4
   [s:(Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)]
   (Cons((k.transport),Var(2) Var(1)) Type(Var(1))[x:Type(Var(1))]#1#6#5#1#3)
   (Cons((k.path_universe),Var(1) Var(1) Var(2))
    #7#5#4
     Proj((k.equiv_fun),
     (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
     (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1))
   #1
   (Proj((k.eissect),
    (Cons((k.isequiv_equiv_path),Var(1) Var(1) Var(1) Var(2)) #7#5#4))
    #1))
type=
 (Ind((k.paths),0,Var(1))
  #4
   ([s:(Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)]
    (Cons((k.transport),Var(2) Var(1))
     Type(Var(1))[x:Type(Var(1))]#1#6#5#1#3)
    (Cons((k.path_universe),Var(1) Var(1) Var(2))
     #7#5#4
      Proj((k.equiv_fun),
      (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
      (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1)))
   ([s:(Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)]
    (Cons((k.transport),Var(2) Var(1))
     Type(Var(1))[x:Type(Var(1))]#1#6#5#1#3)
    #1))

arg 6=
 (Cons((k.transport_path_universe_equiv_path),Var(1) Var(1) Var(2))
  #7#5#4#1#2)
type=
 (Ind((k.paths),0,Var(1))
  (Ind((k.paths),0,Var(1))
   #4
    (Cons((k.transport),Var(2) Var(1))
     Type(Var(1))[X:Type(Var(1))]#1#5#4
      (Cons((k.path_universe),Var(1) Var(1) Var(2))
       #7#5#4
        Proj((k.equiv_fun),
        (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
        (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1))
      #2)
    (Proj((k.equiv_fun),
     (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
     #2))
   (Cons((k.transport_path_universe),Var(1) Var(1) Var(2) Var(1))
    #7#5#4
     Proj((k.equiv_fun),
     (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
     (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1)
     #2)
   (Cons((k.ap),Var(2) Var(1))
    (Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)#4
     [s:(Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)]
     (Cons((k.transport),Var(2) Var(1))
      Type(Var(1))[x:Type(Var(1))]#1#6#5#1#3)
     (Proj((k.equiv_inv),
      (Cons((k.isequiv_equiv_path),Var(1) Var(1) Var(1) Var(2)) #7#5#4))
      (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))#1
     (Proj((k.eissect),
      (Cons((k.isequiv_equiv_path),Var(1) Var(1) Var(1) Var(2)) #7#5#4))
      #1)))


====== type error ====
(Ind((k.paths),0,Var(1))
 (Ind((k.paths),0,Var(1))
  #4
   (Cons((k.transport),Var(2) Var(1))
    Type(Var(1))[X:Type(Var(1))]#1#5#4
     (Cons((k.path_universe),Var(1) Var(1) Var(2))
      #7#5#4
       Proj((k.equiv_fun),
       (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
       (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1))
     #2)
   (Proj((k.equiv_fun),
    (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
    #2))
  (Cons((k.transport_path_universe),Var(1) Var(1) Var(2) Var(1))
   #7#5#4
    Proj((k.equiv_fun),
    (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
    (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1)
    #2)
  (Cons((k.ap),Var(2) Var(1))
   (Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)#4
    [s:(Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)]
    (Cons((k.transport),Var(2) Var(1))
     Type(Var(1))[x:Type(Var(1))]#1#6#5#1#3)
    (Proj((k.equiv_inv),
     (Cons((k.isequiv_equiv_path),Var(1) Var(1) Var(1) Var(2)) #7#5#4))
     (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))#1
    (Proj((k.eissect),
     (Cons((k.isequiv_equiv_path),Var(1) Var(1) Var(1) Var(2)) #7#5#4))
     #1)))
is not convertible with
(Ind((k.paths),0,Var(1))
 (Ind((k.paths),0,Var(1))
  #4
   (Cons((k.transport),Var(2) Var(1))
    Type(Var(1))[X:Type(Var(1))]#1#5#4
     (Cons((k.path_universe),Var(1) Var(1) Var(2))
      #7#5#4
       Proj((k.equiv_fun),
       (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
       Proj((k.equiv_isequiv),
       (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1)))#2)
   (Proj((k.equiv_fun),
    (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
    #2))
  (Cons((k.transport_path_universe),Var(0) Var(1) Var(2) Var(1))
   #7#5#4
    Proj((k.equiv_fun),
    (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
    Proj((k.equiv_isequiv),
    (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))#2)
  (Cons((k.ap),Var(2) Var(1))
   (Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)#4
    [s:(Ind((k.paths),0,Var(2)) Type(Var(1))#5#4)]
    (Cons((k.transport),Var(2) Var(1))
     Type(Var(1))[x:Type(Var(1))]#1#6#5#1#3)
    (Cons((k.path_universe),Var(1) Var(1) Var(2))
     #7#5#4
      Proj((k.equiv_fun),
      (Cons((k.equiv_path),Var(1) Var(1) Var(2) Var(1)) #5#4#1))
      (Cons((k.isequiv_path),Var(1) Var(1) Var(1) Var(1) Var(2)) #5#4#1))
    #1
    (Proj((k.eissect),
     (Cons((k.isequiv_equiv_path),Var(1) Var(1) Var(1) Var(2)) #7#5#4))
     #1)))
====== universes ====
Var(1) = Var(0)
Var(0) < Var(2)
Set <= Var(2)
    <= Var(1)
    <= Var(0)
Prop < Set

Fatal Error: Type error: CantApplyBadType at argument 6
```

**Review by JasonGross — APPROVED** [review]

**Comment by ejgallego** [comment]
Note that the printer doesn't know the terminal width, so please open an issue if you would like `coqtop` `coqchk` to detect it and print accordingly.

**Comment by maximedenes** [comment]
Success here: https://travis-ci.org/ejgallego/coq/builds/350449899

**Pull request merged**